src/Microsoft.AspNet.Server.Kestrel/Http/MemoryPoolBlock2.cs

```diff
@@ -0,0 +1,551 @@
+                        return block.Array[index];
+                    }
+                    else if (block.Next == null)
+                    {
+                        return -1;
+                    }
+                    else
+                    {
+                        block = block.Next;
+                        index = block.Start;
+                    }
+                }
+            }
+
+            public int Seek(int char0)
+            {
+                if (IsDefault)
+                {
+                    return -1;
+                }
+
+                var byte0 = (byte)char0;
+                var vectorStride = Vector<byte>.Count;
+                var ch0Vector = new Vector<byte>(byte0);
+
+                var block = _block;
+                var index = _index;
+                var array = block.Array;
+                while (true)
+                {
+                    while (block.End == index)
+                    {
+                        if (block.Next == null)
+                        {
+                            _block = block;
+                            _index = index;
+                            return -1;
+                        }
+                        block = block.Next;
+                        index = block.Start;
+                        array = block.Array;
+                    }
+                    while (block.End != index)
+                    {
+                        var following = block.End - index;
+                        if (following >= vectorStride)
+                        {
+                            var data = new Vector<byte>(array, index);
+                            var ch0Equals = Vector.Equals(data, ch0Vector);
+                            var ch0Count = Vector.Dot(ch0Equals, _dotCount);
+
+                            if (ch0Count == 0)
+                            {
+                                index += vectorStride;
+                                continue;
+                            }
+                            else if (ch0Count == 1)
+                            {
+                                _block = block;
+                                _index = index + Vector.Dot(ch0Equals, _dotIndex);
+                                return char0;
+                            }
+                            else
+                            {
+                                following = vectorStride;
+                            }
+                        }
+                        for (; following != 0; following--, index++)
+                        {
+                            if (block.Array[index] == byte0)
+                            {
+                                _block = block;
+                                _index = index;
+                                return char0;
+                            }
+                        }
+                    }
+                }
+            }
+
+            public int Seek(int char0, int char1)
```

Comment:
Added comments to all the other mem pool types... I'll move the iterator to it's own file and add comments. Probably rename it to Cursor instead of iterator as it goes...

But yes, this is seeking to char0 or char1, whichever comes first, and returns char0, char1, or -1 based on what it found (-1 if it found neither)